### PR TITLE
Qr Reader

### PR DESCRIPTION
**Pull request by MaximeZmt**
# Qr Reader
### This is related to issue https://github.com/MaximeZmt/SDP_2022-Vibester/issues/221
This Pull Requests contains:

- UserSearchActivity: Add of a button to go to QR Code Scan Activity
- QrCodeActivity: QR code scanner that add friends

In addition the code is tested with a coverage of: [64.8]%

</img></img>

### Notes for reviewers
-  Low coverage because impossible (very difficult) to mock the camera

### CODE REVIEW

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        supportActionBar?.hide()
+        binding = ActivityQrScanningBinding.inflate(layoutInflater)
+        val view = binding.root
+        setContentView(view)
+
+        val extras = intent.extras
+        if (extras != null) {
+            uidList = extras.get("uidList") as ArrayList<String>
+            isTest = extras.getBoolean("isTest", false)
+        }else{
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:56]:
```suggestion
        } else {
```

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            askForCameraPermission()
+        } else {
+            setupControls()
+        }
+
+        val aniSlide: Animation =
+            AnimationUtils.loadAnimation(this@QrScanningActivity, R.anim.qr_anim)
+        binding.barcodeLine.startAnimation(aniSlide)
+
+        if(isTest){
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:76]:
```suggestion
        if (isTest) {
```

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    private lateinit var barcodeDetector: BarcodeDetector
+    private lateinit var binding: ActivityQrScanningBinding
+
+    // Value
+    private var scannedValue = ""
+    var uidList: ArrayList<String> = ArrayList()
+    val usersRepo: DataGetter = DataGetter()
+    var isTest: Boolean = false
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
[the remark this answers is not in the document]
Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:44]: will be a good practice to break onCreate down into smaller functions.

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            }
+
+            @SuppressLint("MissingPermission")
+            override fun surfaceChanged(
+                holder: SurfaceHolder,
+                format: Int,
+                width: Int,
+                height: Int
+            ) {
```
[the remark this answers is not in the document]
Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:110]: the body of surfaceChanged and surfaceCreated are identical, refactor may also boost the coverage

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        if (requestCode == requestCodeCameraPermission && grantResults.isNotEmpty()) {
+            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // If permission are granted restart the activity to make sure they are taken into account
+                val intentNew =  Intent(this, QrScanningActivity::class.java)
+                intentNew.putExtra("uidList", uidList)
+                startActivity(intentNew)
+                finish()
+            } else {
+                // Camera permission not granted, come back to previous activity
+                Toast.makeText(applicationContext, "Camera Permission should be granted for that feature", Toast.LENGTH_LONG).show()
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:172]: can use string.xml

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            }
+
+            override fun surfaceDestroyed(holder: SurfaceHolder) {
+                camera.stop()
+            }
+        })
+
+        barcodeDetector.setProcessor(object : Detector.Processor<Barcode> {
+            override fun release() {
+                Toast.makeText(applicationContext, "Scanner has been closed", Toast.LENGTH_SHORT).show()
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:125]: can use string.xml

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                startActivity(returnInt)
+                finish()
+            }
+        }
+    }
+
+    /*
+     * This section is related to the handler when code is detected
+     */
+
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:183]:
```suggestion
```

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            runOnUiThread {
+                camera.stop()
+                val endInt = Intent(this@QrScanningActivity, SearchUserActivity::class.java)
+                if (isTest || scannedValue in uidList){
+                    if (!isTest) {
+                        usersRepo.setSubFieldValue(FireBaseAuthenticator.getCurrentUID(),"friends", scannedValue, true)
+                    }
+                    Toast.makeText(this@QrScanningActivity, "Congratulations, you have a new friends", Toast.LENGTH_SHORT).show()
+                    endInt.putExtra("isSuccess", true)
+                }else{
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:199]:
```suggestion
                } else {
```

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                scannedValue = barcodes.valueAt(0).rawValue
+            }
+            runOnUiThread {
+                camera.stop()
+                val endInt = Intent(this@QrScanningActivity, SearchUserActivity::class.java)
+                if (isTest || scannedValue in uidList){
+                    if (!isTest) {
+                        usersRepo.setSubFieldValue(FireBaseAuthenticator.getCurrentUID(),"friends", scannedValue, true)
+                    }
+                    Toast.makeText(this@QrScanningActivity, "Congratulations, you have a new friends", Toast.LENGTH_SHORT).show()
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:197]:
```suggestion
                    Toast.makeText(this@QrScanningActivity, "Congratulations, you have a new friend", Toast.LENGTH_SHORT).show()
```

Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:197]: string.xml

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt [hunk 10]
```diff
@@ -136,7 +136,9 @@ class DataGetter @Inject constructor() {
      * @param searchInput search text inputed by user
      * @param callback function to call with found users by username
      */
-    fun searchByField(field: String, searchInput: String, callback:(ArrayList<User>) -> Unit) {
+    fun searchByField(field: String, searchInput: String, callback:(ArrayList<User>) -> Unit, callback2 :(ArrayList<String>) -> Unit) {
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:139]: don't forget to update the function documentation.

Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:139]: and maybe give callback and callback2 more meaningful names

**Review by jiabaow — COMMENTED** [review]
Nice work! I left some comments concerning breaking down large functions into multiple smaller functions and duplications.

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+class QrScanningActivity : AppCompatActivity() {
+    // Const
+    private val requestCodeCameraPermission = 1001
+
+    // API
+    private lateinit var camera: CameraSource
+    private lateinit var barcodeDetector: BarcodeDetector
+    private lateinit var binding: ActivityQrScanningBinding
+
+    // Value
```
Comment (kamilababayeva) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:38]: I believe those comments are excessive

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+                val intentNew =  Intent(this, QrScanningActivity::class.java)
+                intentNew.putExtra("uidList", uidList)
+                startActivity(intentNew)
+                finish()
+            } else {
+                // Camera permission not granted, come back to previous activity
+                Toast.makeText(applicationContext, "Camera Permission should be granted for that feature", Toast.LENGTH_LONG).show()
+                val returnInt = Intent(this@QrScanningActivity, SearchUserActivity::class.java)
+                startActivity(returnInt)
+                finish()
```
Comment (kamilababayeva) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:175]: You have end and start activity in the beginng of the file too. Maybe create a function fo that

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+        if (isTest || barcodes.size() == 1) {
+            if (!isTest) {
+                scannedValue = barcodes.valueAt(0).rawValue
+            }
+            runOnUiThread {
+                camera.stop()
+                val endInt = Intent(this@QrScanningActivity, SearchUserActivity::class.java)
+                if (isTest || scannedValue in uidList){
+                    if (!isTest) {
+                        usersRepo.setSubFieldValue(FireBaseAuthenticator.getCurrentUID(),"friends", scannedValue, true)
```
Comment (kamilababayeva) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:195]: do you wanna try and mock the usersRepo?

File: app/src/main/java/ch/sdp/vibester/activity/SearchUserActivity.kt [hunk 14]
```diff
@@ -73,12 +92,19 @@ class SearchUserActivity : AppCompatActivity() {
         userProfileAdapter.notifyDataSetChanged()
     }
 
+    /**
+     * Callback to update users in adapter during search
+     */
+    private fun setUserInAdapter2(users: ArrayList<String> = ArrayList()) {
+        uidList = ArrayList(users)
```
Comment (kamilababayeva) [inline app/src/main/java/ch/sdp/vibester/activity/SearchUserActivity.kt:99]: why do you need to wrap users in ArrayList? They are already, no?

**Review by kamilababayeva — COMMENTED** [review]
I have did a partial review. I wanna give it a second look later! thank you for the changes! The feature is cool and the moving bar in an old way is the best.

File: app/src/androidTest/java/ch/sdp/vibester/activity/QrScanningActivityTest.kt [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+import org.junit.runner.RunWith
+
+
+@RunWith(AndroidJUnit4::class)
+@HiltAndroidTest
+class QrScanningActivityTest {
+    @get:Rule
+    var hiltRule = HiltAndroidRule(this)
+
+    @get:Rule var permissionRule = GrantPermissionRule.grant(android.Manifest.permission.CAMERA)
```
Comment (kamilababayeva) [inline app/src/androidTest/java/ch/sdp/vibester/activity/QrScanningActivityTest.kt:38]: when we have more than one rule, we specify the order

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt (hunk 13 above)
Comment (MaximeZmt) [reply app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:195]: Maybe not this week, already took so much time to do it. But I can take care of that next week. Depends which task I am taking

File: app/src/main/java/ch/sdp/vibester/activity/SearchUserActivity.kt (hunk 14 above)
Comment (MaximeZmt) [reply app/src/main/java/ch/sdp/vibester/activity/SearchUserActivity.kt:99]: Here this is only the uid. Do you think it is better to do another way ?

**Review by kamilababayeva — APPROVED** [review]
Looks good to me! thanks for a nice feature!

File: app/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -17,8 +17,14 @@
     <string name="authentication_placeholderEmptyScore">Error</string>
     <string name="authentication_noActionDoneYet">Either log in with your Google account or fill in credentials to create an account or log in</string>
 
-    <!-- OLD String-->
+    <!-- QrScanning -->
+    <string name="qrScanning_scannerClosed">Scanner has been closed</string>
+    <string name="qrScanning_cameraError">Camera Permission should be granted for that feature</string>
+    <string name="qrScanning_noExistUid">Error not an existing uid</string>
+    <string name="qrScanning_newFriend">Congratulations, you have a new friends</string>
```
Comment (jiabaow) [inline app/src/main/res/values/strings.xml:24]:
```suggestion
    <string name="qrScanning_newFriend">Congratulations, you have a new friend</string>
```

**Review by jiabaow — APPROVED** [review]
LGTM but don't forget to change the string.xml

**Pull request merged**